Fix StreamService.User(): parse very long friend ID lists properly

mickey24 opened the pull request:
Fixes #73 by replacing `bufio.Scanner` with `bufio.Reader` for parsing streaming data.

`bufio.Scanner` has a line length limit (65,536 bytes by default), which will cause a problem when a user has a large number of friends (say 7,000) and Twitter sends the list of all friend IDs right after establishing User Stream connection. `bufio.Reader` doesn't have the limit and the same problem won't happen.

dougnukem commented on twitter/streams.go:
```diff
@@ -230,21 +231,38 @@ func (s *Stream) retry(req *http.Request, expBackOff backoff.BackOff, aggExpBack
-	// A bufio.Scanner steps through 'tokens' of data on each Scan() using a
-	// SplitFunc. SplitFunc tokenizes input bytes to return the number of bytes
-	// to advance, the token slice of bytes, and any errors.
-	scanner := bufio.NewScanner(body)
```
Another option is to just increase the `bufio` max token size:
- https://golang.org/src/bufio/scan.go?s=3594:3631#L69

Default is 64k, can bump it up to 512k (we found that worked for us).
```go
	scanner := bufio.NewScanner(body)
 +	scanner.Buffer(nil, 512*1024)
```

mickey24 commented:
Thanks for the suggestion! Yes, that could be a simpler option, but I'm afraid the same issue would happen again for users with around >60k friends.

If we want to just increase the max token size, I think we should at least document that go-twitter's User Stream feature has a limitation for users with a large number of friends, or check bufio.Scanner.Err() after the message generation loop and propagate bufio.ErrTooLong to callers. The current implementation silently gets stuck into User Stream connection/disconnection loop and it's hard for users to find what's going wrong.

What do you think?

dougnukem commented:
@mickey24 I think you're right solving it this way, and if we wanted to solve with a larger max buffer we should propagate an error if it goes above it.

Couple thoughts related to this.
- Twitter user stream caps it at some maximum follower size (and then just returns the first 10k or so, I forget what the limit is but we've hit that before)

It might be better to set the stream to "delimited" messages so that we could pre-allocate the buffer when parsing a message rather than relying on a `\r\n`

- https://dev.twitter.com/streaming/overview/processing

> By passing delimited=length when connecting to a stream (note that the value is the literal string length, not a number) each message will be preceded by a string representation of a base-10 integer indicating the length of the message in bytes. Note that this is independent of, and does not affect any chunked transfer encoding. Clients may use these length delimiters to more efficiently copy chunks of text off of the incoming stream, rather than having to parse message text for \r\n tokens.

dougnukem commented on twitter/streams.go:
```diff
@@ -230,21 +231,38 @@ func (s *Stream) retry(req *http.Request, expBackOff backoff.BackOff, aggExpBack
+			}
+			// Otherwise, the line is not the end of a streaming message, so we append
+			// the line to the buffer and continue to scan lines.
+			buf = append(buf, line...)
```
the problem here is that we're creating copies of `[]byte` which could get expensive with a very active twitter stream.

dougnukem replied:
might want to use a `bytes.Buffer` `https://golang.org/pkg/bytes/` (or that can be optimized later)

mickey24 replied:
Thanks for pointing this out. Changed the code to use bytes.Buffer. Could you take another look?

mickey24 commented:
Thanks, changed the code to use bytes.Buffer.

> Couple thoughts related to this.
> - Twitter user stream caps it at some maximum follower size (and then just returns the first 10k or so, I forget what the limit is but we've hit that before)

Looks like the 10k limit is mentioned in [section "Too many follows (warning)"](https://dev.twitter.com/streaming/overview/messages-types#too_many_follows), but it doesn't seem to address whether the same limit will be applied for the friends ID list of User Stream as well. I don't have a Twitter account with >10k friends so I can't test whether bufio.Scanner with a larger max buffer size will solve the issue completely.

> It might be better to set the stream to "delimited" messages so that we could pre-allocate the buffer when parsing a message rather than relying on a `\r\n`

I think this is a reasonable option and the code could be cleaner. If you think this is necessary, I can try to implement it and file a separate pull request after testing with my accounts.

dougnukem commented:
@mickey24 I think the API should probably support both ways and we can implement a new way to support delimited messages in a different PR.

This lgtm for now

dougnukem commented:
@dghubble - if you want to take a look at this or have any thoughts for merging it (I dont have merge access).

dghubble commented:
This changes the tokenization code and deletes the tests for it. :(

mickey24 commented:
@dghubble Thanks, that's a fair-enough point. I've refactored the stream scan logic and added some tests. I hope the important cases in old tests (e.g. reaching EOF without "\r\n") are addressed in the new tests. Could you take another look?

dghubble commented:
Looks good, thanks! Seems like we might be able to just use the bufio.Reader instead of having a buffered reader and a bytes.Buffer, but that can be investigated later.

mickey24 commented:
Great, thank you very much for your review!
